### PR TITLE
WIP Multidict c ext

### PR DESCRIPTION
**Pull request by iemelyanov**

### CODE REVIEW

File: multidict/_multidict_c.c [hunk 1]
```diff
@@ -0,0 +1,854 @@
...
+_multidict_getone(_MultiDictObject *self, PyObject *key, PyObject *_default)
+{
+    PyObject *val = pair_list_get_one(self->impl, key);
+
+    if (val == NULL &&
+        PyErr_ExceptionMatches(PyExc_KeyError) &&
+        _default != NULL)
+    {
+        PyErr_Clear();
+        return _default;
```
Comment (asvetlov) [inline multidict/_multidict_c.c]: Incref `_default`

File: multidict/_multidict_c.c [hunk 2]
```diff
@@ -0,0 +1,854 @@
...
+
+/******************** Base Methods ********************/
+
+/* Forward declaration */
+static PyObject *multidict_items(_MultiDictObject *self);
+
+static INLINE PyObject *
+_multidict_getone(_MultiDictObject *self, PyObject *key, PyObject *_default)
+{
+    PyObject *val = pair_list_get_one(self->impl, key);
```
Comment (asvetlov) [inline multidict/_multidict_c.c]: Ideally, we should avoid setting an exception if not found and clearing the exception on the next step.

File: multidict/_multidict_c.c [hunk 3]
```diff
@@ -0,0 +1,854 @@
...
+            return -1;
+        }
+
+        key   = PyTuple_GET_ITEM(item, 0);
+        value = PyTuple_GET_ITEM(item, 0);
+        pair_list_add(self->impl, key, value);
+
+        Py_DECREF(item);
+    }
+
```
Comment (asvetlov) [inline multidict/_multidict_c.c]: `PyIter_Next()` can return `NULL` **and** set an exception if an error occured.
Should return non-null if `PyErr_Occured()` returns non-zero

Comment (asvetlov) [reply multidict/_multidict_c.c]: Please apply the check to all `PyIter_Next()` calls in the file

File: multidict/_multidict_c.c [hunk 4]
```diff
@@ -0,0 +1,854 @@
...
+    Py_DECREF(iter);
+    return 0;
+}
+
+static int
+_multidict_extend(_MultiDictObject *self, PyObject *args, PyObject *kwds,
+                  const char *name, int do_add)
+{
+    // TOOD: Refactoring me. Split on little functions;
```
Comment (asvetlov) [inline multidict/_multidict_c.c]: The function needs a love, I see many TODOs and agree with them

File: multidict/_multidict_c.c [hunk 5]
```diff
@@ -0,0 +1,854 @@
...
+
+    static char *keywords[] = {"key", "default"};
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O|O:getall",
+                                     keywords, &key, &_default))
+    {
+        return NULL;
+    }
+
+    list = pair_list_get_all(self->impl, key);
```
Comment (asvetlov) [inline multidict/_multidict_c.c]: The same problem with setting and clearing `KeyError`

File: multidict/_multidict_c.c [hunk 6]
```diff
@@ -0,0 +1,854 @@
...
+    if (new_multidict == NULL) {
+        return NULL;
+    }
+
+    items = multidict_items(self);
+    if (items == NULL) {
+        goto fail;
+    }
+
+    arg_items = PyTuple_New(1);
```
Comment (asvetlov) [inline multidict/_multidict_c.c]: Implementation looks as slow as possible but we can improve it later

File: multidict/_multidict_c.c [hunk 7]
```diff
@@ -0,0 +1,854 @@
...
+        return NULL;
+    }
+
+    Py_RETURN_NONE;
+}
+
+static PyObject *
+multidict_clear(_MultiDictObject *self)
+{
+    pair_list_clear(self->impl);
```
Comment (asvetlov) [inline multidict/_multidict_c.c]: Check retval

File: multidict/_multidict_c.c (hunk 1 above)
Comment (iemelyanov) [reply multidict/_multidict_c.c]: Done

File: multidict/_multidict_c.c (hunk 3 above)
Comment (iemelyanov) [reply multidict/_multidict_c.c]: Done

File: multidict/_multidict_c.c (hunk 4 above)
Comment (iemelyanov) [reply multidict/_multidict_c.c]: Partially done. Please look new implementation.

File: multidict/_multidict_c.c (hunk 5 above)
Comment (iemelyanov) [reply multidict/_multidict_c.c]: Done if I understand right :)

File: multidict/_multidict_c.c (hunk 6 above)
Comment (iemelyanov) [reply multidict/_multidict_c.c]: In TODO

File: multidict/_multidict_c.c (hunk 7 above)
Comment (iemelyanov) [reply multidict/_multidict_c.c]: Done

**Comment by webknjaz** [comment]
This pull request **introduces 1 alert** when merging f752ccb3bbf39a4382c5763402a54b6eb5d0f4d5 into 3039dcc57b65d4eebda50ec1cffdedc01a205483 - [view on LGTM.com](https://lgtm.com/projects/g/aio-libs/multidict/rev/pr-ce9e70a448f4758f7b4a290282f61ac047482ad4)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by webknjaz** [comment]
This pull request **introduces 1 alert** when merging 734712b6787494767317535fa6382738e451d0f8 into dcc6422e323a2821057616ab03fa1fd21b50def6 - [view on LGTM.com](https://lgtm.com/projects/g/aio-libs/multidict/rev/pr-6ba61d9601339747c661fb04d1789010f4ed867d)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by webknjaz** [comment]
This pull request **introduces 1 alert** when merging 56cc7854b73e07fb15b409fbbccf9976fbe13261 into dcc6422e323a2821057616ab03fa1fd21b50def6 - [view on LGTM.com](https://lgtm.com/projects/g/aio-libs/multidict/rev/pr-ed15fd20efc75d9c98f2dd567b0db76191cad4f5)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by webknjaz** [comment]
This pull request **introduces 1 alert** when merging c353dada2b5542d55aa76acc5324dd9a73af758e into dcc6422e323a2821057616ab03fa1fd21b50def6 - [view on LGTM.com](https://lgtm.com/projects/g/aio-libs/multidict/rev/pr-58505ccc6bc166f943c2c8df63ee300498a06e3b)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by asvetlov** [comment]
@webknjaz the alert is annoying, isn't it?
`flake8` prevents unused imports anyway, why should we see a message for any commit for work-in-progress PR?

**Comment by webknjaz** [comment]
I've deactivated comments.

File: multidict/_multidict_c.c [hunk 8]
```diff
@@ -0,0 +1,1617 @@
...
+                NULL
+            );
+            Py_DECREF(item);
+            Py_DECREF(iter);
+            return -1;
+        }
+
+        key   = PyTuple_GET_ITEM(item, 0);
+        value = PyTuple_GET_ITEM(item, 1);
+        pair_list_add(self->impl, key, value);
```
Comment (asvetlov) [inline multidict/_multidict_c.c]: check return value for error

File: multidict/_multidict_c.c [hunk 9]
```diff
@@ -0,0 +1,1617 @@
...
+        }
+        if (arg_items == NULL) {
+            return -1;
+        }
+    } else {
+        arg_items = arg;
+        Py_INCREF(arg_items);
+    }
+
+    if (kwds && kwds != Py_None) {
```
Comment (asvetlov) [inline multidict/_multidict_c.c]: How `kwds` can be `None`?
I thought only `NULL` or dict can happen.

Comment (asvetlov) [reply multidict/_multidict_c.c]: Please call [`PyArg_ValidateKeywordArguments()`](https://docs.python.org/3/c-api/arg.html#c.PyArg_ValidateKeywordArguments) before usage.

File: multidict/_multidict_c.c [hunk 10]
```diff
@@ -0,0 +1,1617 @@
...
+
+static INLINE int
+_multidict_extend_with_kwds(MultiDictObject *self, PyObject *kwds,
+                                 const char *name, int do_add)
+{
+    PyObject *arg = NULL;
+
+    int err = 0;
+
+    arg = PyDict_Items(kwds);
```
Comment (asvetlov) [inline multidict/_multidict_c.c]: Please call PyArg_ValidateKeywordArguments() before usage.

File: multidict/_multidict_c.c [hunk 11]
```diff
@@ -0,0 +1,1617 @@
...
+    // TODO: "Implementation looks as slow as possible ..."
+    arg_items = PyTuple_New(1);
+    if (arg_items == NULL) {
+        goto fail;
+    }
+
+    PyTuple_SET_ITEM(arg_items, 0, items);
+
+    if (_multidict_extend(
+        new_multidict, arg_items, Py_None, "copy", 1) < 0)
```
Comment (asvetlov) [inline multidict/_multidict_c.c]: `Py_None` -> `NULL`

File: multidict/_multidict_c.c [hunk 12]
```diff
@@ -0,0 +1,1617 @@
...
+
+static INLINE PyObject *
+_multidict_copy(MultiDictObject *self, PyTypeObject *multidict_tp_object)
+{
+    MultiDictObject *new_multidict = NULL;
+
+    PyObject *arg_items = NULL,
+             *items     = NULL;
+
+    new_multidict = PyObject_GC_New(MultiDictObject, multidict_tp_object);
```
Comment (asvetlov) [inline multidict/_multidict_c.c]: Forgot to call `PyObject_GC_TRACK()` maybe?
Or just `new_multidict = multidict_tp_object->tp_new()`?

File: multidict/_multidict_c.c [hunk 13]
```diff
@@ -0,0 +1,1617 @@
...
+    multidict_methods,                               /* tp_methods */
+    0,                                               /* tp_members */
+    0,                                               /* tp_getset */
+    0,                                               /* tp_base */
+    0,                                               /* tp_dict */
+    0,                                               /* tp_descr_get */
+    0,                                               /* tp_descr_set */
+    0,                                               /* tp_dictoffset */
+    (initproc)multidict_tp_init,                     /* tp_init */
+    0,                                               /* tp_alloc */
```
Comment (asvetlov) [inline multidict/_multidict_c.c]: Python dict has the slot filled with `PyType_GenericAlloc`.
Can we do the same?

File: multidict/_multidict_c.c (hunk 12 above)
Comment (iemelyanov) [reply multidict/_multidict_c.c]: Switched to PyType_GenericNew.

File: multidict/_multidict_c.c (hunk 13 above)
Comment (iemelyanov) [reply multidict/_multidict_c.c]: Yes. Done.

File: multidict/_multidict_c.c (hunk 8 above)
Comment (iemelyanov) [reply multidict/_multidict_c.c]: Done.

File: multidict/_compat.py [hunk 14]
```diff
@@ -6,4 +6,4 @@
 
 PYPY = platform.python_implementation() == 'PyPy'
 
-USE_CYTHON_EXTENSIONS = USE_CYTHON = not NO_EXTENSIONS and not PYPY
+USE_C_EXTENSIONS = not NO_EXTENSIONS and not PYPY
```
Comment (webknjaz) [inline multidict/_compat.py]: Public API change, needs a deprecation warning.

File: tests/gen_pickles.py [hunk 15]
```diff
@@ -19,8 +19,9 @@ def write(name, proto):
 
 
 def generate():
-    if not USE_CYTHON:
-        raise RuntimeError("Cython is required")
+    if not USE_C_EXTENSIONS:
+        # TODO: mb this useless and should be refactored
+        raise RuntimeError("C is required")
```
Comment (webknjaz) [inline tests/gen_pickles.py]: C? Does a normal user need to know what that is? GCC I guess

Comment (iemelyanov) [reply tests/gen_pickles.py]: Good point. I will change it.

File: multidict/_compat.py (hunk 14 above)
Comment (iemelyanov) [reply multidict/_compat.py]: What I should do? Can you show an example?

**Comment by iemelyanov** [comment]
> Most of this change is just renaming the `USE_CYTHON_EXTENSIONS` variable. It makes it impossible to identify "real" changes in this diff. Please do this as a separate PR.

Ok, I will revert this branch to 2929ecb9316127ecff46eca7cbf03fde69bd503e and move "renaming consts" to new branch.

File: multidict/_compat.py (hunk 14 above)
Comment (webknjaz) [reply multidict/_compat.py]: one of the ways is to declare `__getattribute__(name)` function but that'd work under Python 3.7 only. the other way is to have a hack with replacing the module namespace with a class instance.

**Comment by asvetlov** [comment]
Superseded by #394

**Pull request closed without merging**